File: Project/PluginDev/lite/base/src/main/java/com/larry/lite/base/LiteStrategy.java
package com.larry.lite.base;

public class LiteStrategy {
    public static final int PERIODICITY_PARAM_HOUR = 3600000;
    public static final int PERIODICITY_PARAM_HALF_DAY = 43200000;
    public static final int PERIODICITY_PARAM_DAY = 86400000;
    public static final int PERIODICITY_PARAM_WEEK = 604800000;
    public static final int KEYEVENT_PARAM_START = 1;
    public static final int KEYEVENT_PARAM_UPGRADE = 2;
    public static final int KEYEVENT_PARAM_BACKGROUND = 3;
    public final LiteLaunch mode;
    public final LiteNetworkType networkLimit;
    public final int modeExtra;
    public final int limit;

    private LiteStrategy(LiteLaunch m, LiteNetworkType nt, int mp, int l) {
        this.mode = m;
        this.networkLimit = nt;
        this.modeExtra = mp;
        this.limit = l;
    }

    public boolean equals(Object o) {
        if (this == o) {
            return true;
        } else if (o != null && this.getClass() == o.getClass()) {
            LiteStrategy that = (LiteStrategy) o;
            return this.modeExtra != that.modeExtra
                    ? false
                    : (this.limit != that.limit
                            ? false
                            : (this.mode != that.mode ? false : this.networkLimit == that.networkLimit));
        } else {
            return false;
        }
    }

    public int hashCode() {
        int result = this.mode.hashCode();
        result = 31 * result + this.networkLimit.hashCode();
        result = 31 * result + this.modeExtra;
        result = 31 * result + this.limit;
        return result;
    }

    public static LiteStrategy.Builder newBuilder() {
        return new LiteStrategy.Builder();
    }

    public String toString() {
        return "{limit=" + this.limit + ", mode=" + this.mode + ", networkLimit=" + this.networkLimit + ", modeExtra="
                + this.modeExtra + '}';
    }

    public static class Builder {
        public LiteLaunch mode;
        public LiteNetworkType networkLimit;
        public int modeExtra;
        public int limit;

        Builder() {
            this.mode = LiteLaunch.Periodicity;
            this.modeExtra = 604800000;
            this.networkLimit = LiteNetworkType.WIFI;
            this.limit = 0;
        }

        public LiteStrategy.Builder setLimit(int limit) {
            this.limit = limit;
            return this;
        }

        public LiteStrategy.Builder setMode(LiteLaunch mode) {
            this.mode = mode;
            return this;
        }

        public LiteStrategy.Builder setModeExtra(int extra) {
            this.modeExtra = extra;
            return this;
        }

        public LiteStrategy.Builder setNetworkLimit(LiteNetworkType networkLimit) {
            this.networkLimit = networkLimit;
            return this;
        }

        public LiteStrategy build() {
            if (this.mode != null && this.networkLimit != null) {
                return new LiteStrategy(this.mode, this.networkLimit, this.modeExtra, this.limit);
            } else {
                throw new NullPointerException("mode or networkLimit is null");
            }
        }
    }
}

File: Project/PluginDev/lite/base/src/main/java/com/larry/lite/base/LiteStrategy.kt
package com.larry.lite.base

class LiteStrategy private constructor(val mode: LiteLaunch, val networkLimit: LiteNetworkType, val modeExtra: Int, val limit: Int) {

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        } else if (o != null && this.javaClass == o.javaClass) {
            val that = o as LiteStrategy?
            return if (this.modeExtra != that!!.modeExtra)
                false
            else
                if (this.limit != that.limit)
                    false
                else
                    if (this.mode !== that.mode) false else this.networkLimit === that.networkLimit
        } else {
            return false
        }
    }

    override fun hashCode(): Int {
        var result = this.mode.hashCode()
        result = 31 * result + this.networkLimit.hashCode()
        result = 31 * result + this.modeExtra
        result = 31 * result + this.limit
        return result
    }

    override fun toString(): String {
        return "{limit=" + this.limit + ", mode=" + this.mode + ", networkLimit=" + this.networkLimit + ", modeExtra=" + this.modeExtra + "}"
    }

    class Builder internal constructor() {
        var mode: LiteLaunch? = null
        var networkLimit: LiteNetworkType? = null
        var modeExtra: Int = 0
        var limit: Int = 0

        init {
            this.mode = LiteLaunch.Periodicity
            this.modeExtra = 604800000
            this.networkLimit = LiteNetworkType.WIFI
            this.limit = 0
        }

        fun setLimit(limit: Int): LiteStrategy.Builder {
            this.limit = limit
            return this
        }

        fun setMode(mode: LiteLaunch): LiteStrategy.Builder {
            this.mode = mode
            return this
        }

        fun setModeExtra(extra: Int): LiteStrategy.Builder {
            this.modeExtra = extra
            return this
        }

        fun setNetworkLimit(networkLimit: LiteNetworkType): LiteStrategy.Builder {
            this.networkLimit = networkLimit
            return this
        }

        fun build(): LiteStrategy {
            if (this.mode != null && this.networkLimit != null) {
                return LiteStrategy(this.mode!!, this.networkLimit!!, this.modeExtra, this.limit)
            } else {
                throw NullPointerException("mode or networkLimit is null")
            }
        }
    }

    companion object {
        val PERIODICITY_PARAM_HOUR = 3600000
        val PERIODICITY_PARAM_HALF_DAY = 43200000
        val PERIODICITY_PARAM_DAY = 86400000
        val PERIODICITY_PARAM_WEEK = 604800000
        val KEYEVENT_PARAM_START = 1
        val KEYEVENT_PARAM_UPGRADE = 2
        val KEYEVENT_PARAM_BACKGROUND = 3

        fun newBuilder(): LiteStrategy.Builder {
            return LiteStrategy.Builder()
        }
    }
}
